--- xptcinvoke_gcc_x86_unix.cpp	2011-12-20 20:07:13.000000000 -0500
+++ xptcinvoke_gcc_x86_unix.cpp	2014-04-13 20:16:10.330213099 -0400
@@ -82,22 +82,6 @@
   params      = ebp + 0x14
 
 */
-/*
- * Hack for gcc for win32 and os2.  Functions used externally must be
- * explicitly dllexported.
- * Bug 226609
- */
-#if defined(XP_WIN32) || defined(XP_OS2)
-extern "C" {
-    nsresult _NS_InvokeByIndex_P(nsISupports* that, PRUint32 methodIndex,
-                               PRUint32 paramCount, nsXPTCVariant* params);
-    EXPORT_XPCOM_API(nsresult)
-    NS_InvokeByIndex_P(nsISupports* that, PRUint32 methodIndex,
-                     PRUint32 paramCount, nsXPTCVariant* params) { 
-        return _NS_InvokeByIndex_P(that, methodIndex, paramCount, params);
-    }
-}
-#endif
 
 __asm__ (
 	".text\n\t"
@@ -143,7 +127,11 @@
 	"movl  %ebp, %esp\n\t"
 	"popl  %ebp\n\t"
 	"ret\n"
-#if !defined(XP_WIN32) && !defined(XP_OS2) && !defined(XP_MACOSX)
+#if defined(XP_WIN32) || defined(XP_OS2)
+  ".section .drectve\n\t"
+  ".ascii \" -export:NS_InvokeByIndex_P\"\n\t"
+  ".text\n\t"
+#elif !defined(XP_MACOSX)
 	".size " SYMBOL_UNDERSCORE "NS_InvokeByIndex_P, . -" SYMBOL_UNDERSCORE "NS_InvokeByIndex_P\n\t"
 #endif
 );
