--- accessible/src/msaa/nsAccessNodeWrap.h.orig	2014-04-14 11:02:01.270425697 -0400
+++ accessible/src/msaa/nsAccessNodeWrap.h	2014-04-14 11:02:27.941424546 -0400
@@ -58,8 +58,8 @@
 #include "nsIDOMElement.h"
 #include "nsIContent.h"
 #include "nsAccessNode.h"
-#include "OLEIDL.H"
-#include "OLEACC.H"
+#include "oleidl.h"
+#include "oleacc.h"
 #include <winuser.h>
 #ifdef MOZ_CRASHREPORTER
 #include "nsICrashReporter.h"
